File: search_in_rotated_sorted_array.py
class Solution:
    def search(self, nums, target):
        if not nums:
            return -1
        return self.binarySearch(nums, target, 0, len(nums)-1)
    
    def binarySearch(self, nums, target, start, end):
        if end < start:
            return -1
        mid = (start+end)//2
        if nums[mid] == target:
            return mid
        if nums[start] <= target < nums[mid]: # left side is sorted and has target
            return self.binarySearch(nums, target, start, mid-1)
        elif nums[mid] < target <= nums[end]: # right side is sorted and has target
            return self.binarySearch(nums, target, mid+1, end)
        elif nums[mid] > nums[end]: # right side is pivoted
            return self.binarySearch(nums, target, mid+1, end)
        else: # left side is pivoted
            return self.binarySearch(nums, target, start, mid-1)

File: search_in_rotated_sorted_array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        
        int mid, low = 0, high = nums.size() - 1;
        int midEle, cLeft, cRight;
        
        while(low <= high){
            mid = low + (high - low) / 2;
            
            if(nums[mid] == target) return mid;

            //right half is sorted
            if(nums[mid] <= nums[high]){
                if(nums[mid] <= target && target <= nums[high]){
                    low = mid + 1;
                }
                else
                    high = mid - 1;
            }
            
            //left half is sorted
            else{
                if(nums[low] <= target && target <= nums[mid]){
                    high = mid - 1;
                }
                else
                    low = mid + 1;
            }
        }
        return -1;
    }
};